docs/developer-guide/working-with-plugins.md

```diff
@@ -246,4 +246,33 @@ exports.parseForESLint = function(code, options) {
 
 ```
 
+### The AST specification
 
+The AST that custom parsers should create is based on [ESTree](https://github.com/estree/estree) basically. The AST requires some additional properties about detail information of the source code.
+
+#### All nodes:
+
+All nodes must have `range` property.
+
+* `range` (`number[]`) is the array of two numbers. Both numbers are a 0-based index which is the position in the array of source code characters. The first is the start position of the node, the second is the end position of the node. `code.slice(node.range[0], node.range[1])` must be the text of the node. This range does not include spaces which are around the node.
+* `loc` (`SourceLocation`) must not be `null`. [The `loc` property is defined as nullable by ESTree](https://github.com/estree/estree/blob/25834f7247d44d3156030f8e8a2d07644d771fdb/es5.md#node-objects), but ESLint requires this property. On the other hand, `SourceLocation#source` property can be `undefined`. ESLint does not use the `SourceLocation#source` property.
+
+The `parent` property of all nodes must be rewriteable. ESLint set their parent node to the `parent` properties while traversing.
```

Comment:
I feel the second sentence would flow better with the first by reversing the nouns:

> ESLint sets each node's `parent` properties to its parent node while traversing.